lightning/src/chain/keysinterface.rs

```diff
@@ -392,16 +393,60 @@ pub enum Recipient {
 	PhantomNode,
 }
 
+///
+pub trait SharedSecretProduce: Send + Sync {
+	///
+	fn public_key(&self, secp_ctx: &secp256k1::Secp256k1<SignOnly>) -> PublicKey;
+	///
+	fn shared_secret(&self, other: &PublicKey) -> SharedSecret;
+	///
+	fn do_clone(&self) -> Box<dyn SharedSecretProduce>;
+}
+
+///
+pub struct EphemeralSharedSecretProducer {
+	///
+	pub secret_key: SecretKey
+}
+
+impl SharedSecretProduce for EphemeralSharedSecretProducer {
+	fn public_key(&self, secp_ctx: &Secp256k1<SignOnly>) -> PublicKey {
+		PublicKey::from_secret_key(&secp_ctx, &self.secret_key)
+	}
+
+	fn shared_secret(&self, other: &PublicKey) -> SharedSecret {
+		SharedSecret::new(other, &self.secret_key)
+	}
+
+	fn do_clone(&self) -> Box<dyn SharedSecretProduce> {
+		Box::new(Self { secret_key: self.secret_key })
+	}
+}
+
+impl EphemeralSharedSecretProducer {
+	///
+	pub fn new(secret_key: SecretKey) -> Box<dyn SharedSecretProduce> {
+		Box::new( Self { secret_key })
+	}
+}
+
 /// A trait to describe an object which can get user secrets and key material.
 pub trait KeysInterface {
 	/// A type which implements Sign which will be returned by get_channel_signer.
 	type Signer : Sign;
 
-	/// Get node secret key (aka node_id or network_key) based on the provided [`Recipient`].
+	/// A shared secret producer using the node key
+	fn get_shared_secret_producer(&self) -> Box<dyn SharedSecretProduce>;
```

Comment:
Not quite sure I understand the need for this trait here, and super not a fan of box-dyn in our interface.

Reply:
The trait replaces `SecretKey` in several places.  It seems particularly necessary in `DirectionalNoiseState`, so that low-level protocol enum doesn't get exposed to the full `KeysInterface`.

The `Box<dyn>` was expedient for the draft - a generic would propagate into a lot of peer protocol structs.